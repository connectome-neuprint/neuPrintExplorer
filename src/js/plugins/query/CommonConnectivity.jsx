/*
 * Queries common inputs/outputs given list of bodyIds
 */
import React from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import Button from '@mui/material/Button';
import FormControl from '@mui/material/FormControl';
import FormControlLabel from '@mui/material/FormControlLabel';
import Radio from '@mui/material/Radio';
import RadioGroup from '@mui/material/RadioGroup';
import TextField from '@mui/material/TextField';

import NeuronFilterNew, {
  convertToCypher,
  thresholdCypher,
  statusCypher
} from './shared/NeuronFilterNew';
import { getBodyIdForTable } from './shared/pluginhelpers';

const styles = theme => ({
  formControl: {
    margin: theme.spacing(1)
  },
  chips: {
    display: 'flex',
    flexWrap: 'wrap'
  }
});

const pluginName = 'CommonConnectivity';
const pluginAbbrev = 'cc';

const groupBy = (inputJson, key) =>
  inputJson.reduce((accumulator, currentValue) => {
    // name of the common input/output
    const { name, type } = currentValue;
    // first element of the keys array is X_weight where X is the body id of a queried neuron
    let weights = Object.keys(currentValue)[0];
    // in case order of keys changes check that this is true and if not find the correct key
    if (!weights.endsWith('weight')) {
      for (let i = 1; i < Object.keys(currentValue).length; i += 1) {
        if (Object.keys(currentValue)[i].endsWith('weight')) {
          weights = Object.keys(currentValue)[i];
          break;
        }
      }
    }
    (accumulator[currentValue[key]] = accumulator[currentValue[key]] || {})[weights] =
      currentValue[weights];
    accumulator[currentValue[key]].name = name;
    accumulator[currentValue[key]].type = type;
    return accumulator;
  }, {});

class CommonConnectivity extends React.Component {
  static get details() {
    return {
      name: pluginName,
      displayName: 'Common connectivity',
      abbr: pluginAbbrev,
      description:
        'Finds common inputs/outputs for a group of bodies and weights of their connections to these inputs/outputs.',
      visType: 'SimpleTable'
    };
  }

  static getColumnHeaders(query) {
    const { pm: parameters } = query;

    const queryKey = parameters.find_inputs ? 'input' : 'output';

    const columns = [
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} ID`,
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} Instance`,
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} Type`
    ];

    let selectedNeurons = [];
    if (parameters.neuron_ids.length > 0) {
      selectedNeurons = parameters.neuron_ids;
    } else {
      selectedNeurons = parameters.neuron_names;
    }

    const selectedWeightHeadings = selectedNeurons.map(neuron => `${neuron}_weight`);
    selectedWeightHeadings.forEach(neuronWeightHeading => {
      columns.push(neuronWeightHeading);
    });

    columns.push('Total Weight');

    return columns.map(column => ({ name: column, status: true }));
  }

  static fetchParameters(params) {
    const filters = params.filters ? Object.entries(params.filters).map(([filterName, value]) => (
      convertToCypher(filterName, Array.isArray(value) ? value : [value])
    )) : [];
    const conditions = [
      thresholdCypher('pre', params.pre),
      thresholdCypher('post', params.post),
      statusCypher(params.statuses),
      ...filters
    ].filter(condition => condition !== '').join(' AND ');

    const hasConditions = conditions.length > 0 ? 'AND' : '';

    const neuronIds = params.neuron_ids ? params.neuron_ids.join(', ') : '';

    const matchCypher = params.find_inputs ? '(k:Neuron)<-[r:ConnectsTo]-(neuron)' : '(k:Neuron)-[r:ConnectsTo]->(neuron)'

    const cypherQuery = `WITH [${neuronIds}] AS queriedNeurons
MATCH ${matchCypher}
WHERE (k.bodyId IN queriedNeurons ${hasConditions} ${conditions})
WITH k, neuron, r, toString(k.bodyId)+"_weight" AS dynamicWeight
RETURN collect(apoc.map.fromValues(["${params.find_inputs ? "input" : "output"}",
      toString(neuron.bodyId),
      "name",
      neuron.instance,
      "type",
      neuron.type,
      dynamicWeight,
      r.weight])) AS map`;

    return {
      cypherQuery,
      queryString: '/custom/custom?np_explorer=commmon_connectivity'
    };
  }

  static processDownload(response) {
    const {
      find_inputs: findInputs,
      neuron_ids: neuronIds,
      neuron_names: neuronNames
    } = response.params.pm;

    const queryKey = findInputs ? 'input' : 'output';
    const connectionArray = response.result.data[0][0];

    const columnHeaders = [
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} ID`,
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} Instance`,
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} Type`
    ];

    const groupedByInputOrOutputId = groupBy(connectionArray, queryKey);

    let selectedNeurons = [];
    if (neuronIds.length > 0) {
      selectedNeurons = neuronIds;
    } else {
      selectedNeurons = neuronNames;
    }

    const selectedWeightHeadings = selectedNeurons.map(neuron => `${neuron}_weight`);
    selectedWeightHeadings.forEach(neuronWeightHeading => {
      columnHeaders.push(neuronWeightHeading);
    });

    columnHeaders.push('Total Weight');

    const data = [];
    Object.keys(groupedByInputOrOutputId).forEach(inputOrOutput => {
      const singleRow = [
        inputOrOutput,
        groupedByInputOrOutputId[inputOrOutput].name,
        groupedByInputOrOutputId[inputOrOutput].type
      ];

      let totalWeight = 0;
      selectedWeightHeadings.forEach(selectedWeightHeading => {
        const selectedWeightValue =
          groupedByInputOrOutputId[inputOrOutput][selectedWeightHeading] || 0;
        singleRow.push(parseInt(selectedWeightValue, 10));
        totalWeight += parseInt(selectedWeightValue, 10);
      });
      // calculate the total weight for this row.
      singleRow.push(totalWeight);
      data.push(singleRow);
    });

    data.unshift(columnHeaders);
    return data;
  }

  static processResults({ query, apiResponse, actions }) {
    const { pm: parameters } = query;

    const queryKey = parameters.find_inputs ? 'input' : 'output';
    const connectionArray = apiResponse.data[0][0];

    const columns = [
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} ID`,
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} Instance`,
      `${queryKey[0].toUpperCase() + queryKey.substring(1)} Type`
    ];

    const groupedByInputOrOutputId = groupBy(connectionArray, queryKey);

    let selectedNeurons = [];
    if (parameters.neuron_ids.length > 0) {
      selectedNeurons = parameters.neuron_ids;
    } else {
      selectedNeurons = parameters.neuron_names;
    }

    const selectedWeightHeadings = selectedNeurons.map(neuron => `${neuron}_weight`);
    selectedWeightHeadings.forEach(neuronWeightHeading => {
      columns.push(neuronWeightHeading);
    });

    columns.push('Total Weight');

    const data = [];
    Object.keys(groupedByInputOrOutputId).forEach(inputOrOutput => {
      const singleRow = [
        getBodyIdForTable(query.ds, inputOrOutput, actions),
        groupedByInputOrOutputId[inputOrOutput].name,
        groupedByInputOrOutputId[inputOrOutput].type
      ];
      let totalWeight = 0;
      selectedWeightHeadings.forEach(selectedWeightHeading => {
        const selectedWeightValue =
          groupedByInputOrOutputId[inputOrOutput][selectedWeightHeading] || 0;
        singleRow.push(parseInt(selectedWeightValue, 10));
        totalWeight += parseInt(selectedWeightValue, 10);
      });
      // calculate the total weight for this row.
      singleRow.push(totalWeight);
      data.push(singleRow);
    });

    return {
      columns,
      data,
      debug: apiResponse.debug,
      title: `Common ${queryKey}s for ${selectedNeurons} in ${parameters.dataset}`
    };
  }

  constructor(props) {
    super(props);
    this.state = {
      limitNeurons: true,
      status: [],
      pre: 0,
      post: 0,
      filters: {},
      bodyIds: '',
      typeValue: 'input'
    };
  }

  processRequest = () => {
    const { dataSet, submit, actions } = this.props;
    const { limitNeurons, pre, post, status, filters } = this.state;
    const { bodyIds, typeValue } = this.state;

    const parameters = {
      dataset: dataSet,
      statuses: status,
      find_inputs: typeValue !== 'output',
      neuron_ids: bodyIds === '' ? [] : bodyIds.split(','),
      all_segments: !limitNeurons
    };

    if (parameters.neuron_ids.length > 100) {
      actions.metaInfoError(
        `You entered ${
          parameters.neuron_ids.length
        } Neuron IDs. Please limit the list to 100 or less`
      );
      return;
    }

    if (parameters.neuron_ids.length < 1) {
      actions.metaInfoError("You must enter at least one Neuron ID.");
      return;
    }

    if (pre > 0) {
      parameters.pre = pre;
    }

    if (post > 0) {
      parameters.post = post;
    }

    if (Object.keys(filters).length > 0) {
      parameters.filters = filters;
    }

    const query = {
      dataSet,
      plugin: pluginName,
      pluginCode: pluginAbbrev,
      parameters
    };

    submit(query);
  };

  loadNeuronFiltersNew = filters => {
    this.setState({
      filters
    });
  };

  addNeuronBodyIds = event => {
    this.setState({
      bodyIds: event.target.value
    });
  };

  setInputOrOutput = event => {
    this.setState({ typeValue: event.target.value });
  };

  catchReturn = event => {
    // submit request if user presses enter
    if (event.keyCode === 13) {
      event.preventDefault();
      this.processRequest();
    }
  };

  render() {
    const { classes, dataSet, actions, neoServerSettings } = this.props;
    const { bodyIds, typeValue } = this.state;

    return (
      <div>
        <FormControl fullWidth className={classes.formControl}>
          <TextField
            label="Neuron IDs"
            multiline
            fullWidth
            rows={1}
            value={bodyIds}
            name="bodyIds"
            rowsMax={4}
            helperText="Separate IDs with commas. Max 100"
            onChange={this.addNeuronBodyIds}
            onKeyDown={this.catchReturn}
          />
        </FormControl>
        <RadioGroup
          aria-label="Type Of Connections"
          name="type"
          value={typeValue}
          onChange={this.setInputOrOutput}
        >
          <FormControlLabel value="input" control={<Radio color="primary" />} label="Inputs" />
          <FormControlLabel value="output" control={<Radio color="primary" />} label="Outputs" />
        </RadioGroup>
        <NeuronFilterNew
          callback={this.loadNeuronFiltersNew}
          datasetstr={dataSet}
          actions={actions}
          neoServer={neoServerSettings.get('neoServer')}
        />
        <Button variant="contained" color="primary" onClick={this.processRequest}>
          Submit
        </Button>
      </div>
    );
  }
}

CommonConnectivity.propTypes = {
  dataSet: PropTypes.string.isRequired,
  actions: PropTypes.object.isRequired,
  submit: PropTypes.func.isRequired,
  neoServerSettings: PropTypes.object.isRequired,
  classes: PropTypes.object.isRequired
};

export default withStyles(styles, { withTheme: true })(CommonConnectivity);
