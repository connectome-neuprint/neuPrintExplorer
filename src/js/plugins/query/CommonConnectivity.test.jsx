import React from 'react';
import { render, screen } from '../../tests/test-utils';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
// eslint-disable-next-line import/no-unresolved
import '@testing-library/jest-dom';
import CommonConnectivity from './CommonConnectivity';

const { actions, submit } = global;

const styles = {};

const neoServerSettings = {
  get: () => 'http://example.com'
};

const raw = (
  <MemoryRouter>
    <CommonConnectivity
      actions={actions}
      submit={submit}
      dataSet="mb6"
      history={{ push: jest.fn() }}
      classes={styles}
      neoServerSettings={neoServerSettings}
    />
  </MemoryRouter>
);

describe('Common Connectivity Plugin', () => {
  beforeEach(() => {
    submit.mockClear();
    actions.metaInfoError.mockClear();
  });

  describe('has required functions', () => {
    test('name', () => {
      expect(CommonConnectivity.details.name).toBeTruthy();
    });
    test('description', () => {
      expect(CommonConnectivity.details.description).toBeTruthy();
    });
  });

  describe('renders correct defaults', () => {

    it('should render', () => {
      const { asFragment } = render(raw);
      expect(asFragment()).toMatchSnapshot();
    });

    test('bodyIds should be empty', async () => {
      render(raw);
      const bodyIds = screen.getByRole('textbox');
      expect(bodyIds).toHaveValue('');
    });

    test('typeValue should be inputs', async () => {
      render(raw);
      const inputType = screen.getByRole('radio', { name: 'Inputs' });
      expect(inputType).toBeChecked();

      const outputType = screen.getByRole('radio', { name: 'Outputs' });
      expect(outputType).not.toBeChecked();
    });
  });

  describe('submits correctly', () => {
    test('submit button pressed', async () => {
      render(raw);
      const inputArea = await screen.getByRole('textbox');
      await userEvent.type( inputArea, '1234' );
      const submitButton = await screen.getByRole('button', { name: 'Submit' });
      submitButton.click();
      expect(actions.metaInfoError).toHaveBeenCalledTimes(2);
      expect(submit).toHaveBeenCalledTimes(1);
    });

    test('enter key pressed in text field', async () => {
      render(raw);
      const inputArea = await screen.getByRole('textbox');
      await userEvent.type( inputArea, '{enter}' );
      expect(actions.metaInfoError).toHaveBeenCalledTimes(2);
      expect(submit).toHaveBeenCalledTimes(0);
    });
  });
});
