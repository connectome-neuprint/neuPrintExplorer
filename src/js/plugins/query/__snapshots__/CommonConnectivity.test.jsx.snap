// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common Connectivity Plugin renders correct defaults should render 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth CommonConnectivity-formControl-1 css-ytlejw-MuiFormControl-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-2hy7sk-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          Neuron IDs
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-1fqsnzy-MuiInputBase-root-MuiInput-root"
        >
          <textarea
            aria-describedby=":r0:-helper-text"
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-9j73qj-MuiInputBase-input-MuiInput-input"
            id=":r0:"
            name="bodyIds"
            rows="1"
            style="height: 0px; overflow: hidden;"
          />
          <textarea
            aria-hidden="true"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline css-9j73qj-MuiInputBase-input-MuiInput-input"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
            tabindex="-1"
          />
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-sizeMedium css-18y2hez-MuiFormHelperText-root"
          id=":r0:-helper-text"
        >
          Separate IDs with commas. Max 100
        </p>
      </div>
    </div>
    <div
      aria-label="Type Of Connections"
      class="MuiFormGroup-root MuiRadioGroup-root css-1mac7pt-MuiFormGroup-root"
      role="radiogroup"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-19opgx6-MuiFormControlLabel-root"
      >
        <span
          class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary Mui-checked MuiRadio-root MuiRadio-colorPrimary css-13lvt8g-MuiButtonBase-root-MuiRadio-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-j8yymo"
            name="type"
            type="radio"
            value="input"
          />
          <span
            class="css-1qiat4j"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1bz1rr0-MuiSvgIcon-root"
              data-testid="RadioButtonUncheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-z8nmqa-MuiSvgIcon-root"
              data-testid="RadioButtonCheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
              />
            </svg>
          </span>
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-rizt0-MuiTypography-root"
        >
          Inputs
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-19opgx6-MuiFormControlLabel-root"
      >
        <span
          class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-root MuiRadio-colorPrimary css-13lvt8g-MuiButtonBase-root-MuiRadio-root"
        >
          <input
            class="PrivateSwitchBase-input css-j8yymo"
            name="type"
            type="radio"
            value="output"
          />
          <span
            class="css-1qiat4j"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1bz1rr0-MuiSvgIcon-root"
              data-testid="RadioButtonUncheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-o8rdeq-MuiSvgIcon-root"
              data-testid="RadioButtonCheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
              />
            </svg>
          </span>
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-rizt0-MuiTypography-root"
        >
          Outputs
        </span>
      </label>
    </div>
    <div
      class="makeStyles-expandablePanel-4"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1808mag-MuiPaper-root-MuiAccordion-root"
        style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
      >
        <h3
          class="MuiAccordion-heading css-cy7rkm-MuiAccordion-heading"
        >
          <button
            aria-expanded="false"
            class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-1f1pi2l-MuiButtonBase-root-MuiAccordionSummary-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-yfrx4k-MuiAccordionSummary-content"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-1i3tsf3-MuiTypography-root"
              >
                Optional neuron/segment filters
              </h6>
            </span>
            <span
              class="MuiAccordionSummary-expandIconWrapper css-1wqf3nl-MuiAccordionSummary-expandIconWrapper"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="ExpandMoreIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
              </svg>
            </span>
          </button>
        </h3>
        <div
          class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-cwrbtg-MuiCollapse-root"
          style="min-height: 0px;"
        >
          <div
            class="MuiCollapse-wrapper MuiCollapse-vertical css-1x6hinx-MuiCollapse-wrapper"
          >
            <div
              class="MuiCollapse-wrapperInner MuiCollapse-vertical css-1i4ywhz-MuiCollapse-wrapperInner"
            >
              <div
                class="MuiAccordion-region"
                role="region"
              >
                <div
                  class="MuiAccordionDetails-root makeStyles-nopad-5 css-1lj39kh-MuiAccordionDetails-root"
                >
                  <p>
                    Loading...
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-74d805-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Submit
    </button>
  </div>
</DocumentFragment>
`;
